src/Core/CustomMethods.php

```diff
@@ -145,7 +145,15 @@ protected function registerExtraMethodCallback($name, $callback)
      */
     public function hasMethod($method)
     {
-        return method_exists($this, $method) || $this->getExtraMethodConfig($method);
+        $hasMethod = method_exists($this, $method) || $this->getExtraMethodConfig($method);
+
+        if (!$hasMethod) {
```

Comment:
If hasMethod is called on a method that doesn't exist, we're adding some expensive code here. hasMethod returning false could be designed behaviour on a site, and this change risks materially slowing that site down.

I'd treat this with extreme caution. I'm not sure that this is the right fix, and I think that this would be too risky for a late-stage 4.3 change, and too risky for a feature that merely tidies up tests. :-/

Reply:
I'm a little confused as to why this is necessary.

 * Extensible::add_extension() and Extensible::remove_extension() reset the custom methods for the named class and all subclasses, forcing a defineMethods() call the next time getExtraMethodConfig() is called.

Reply:
Yeah ok, fair enough. The background here is this:

* #8465 allows for config for extensions to be overloaded with injector after the extra_methods have been calculated for the original. This was fixing a bug (shown in the unit tests added in that PR) where an overloaded extension was not having its extra methods registered in a key project's unit test suite - it was global state leakage between tests - this came down to the fact that the extra methods for the original Extension class had already been defined in setUpOnce (or similar)
* #8505 fixes a bug in advanced workflow that was caused by the changes to ExtensionTestState in #8465. It was a defensive change, but also added the extra call to reset the extra methods between tests.
* This PR was fixing a seg fault in PHP 5.6 unit tests, and ensuring that the original test added in #8465 would still pass by doing a second call to defineMethods during the hasMethod() check (this is what you're pointing out might be risky)

You're right that add/remove_extension calls reset the extra_methods cache for the object you're applying the extension to. The original bug was from modifying extension class config after those had been generated, so extra methods on the overloaded class were not available.

I've done some testing, and I think we should revert all of these and change the unit test to call `Member::add_extension(SomeExtension::class)` after it changes the config for that extension. This fixes the issue (I'll also pass on the message to the key project that had this problem).

I'm actually not confident that the changes I made are correctly fixing the problem, there's still the odd test state leak happening and it's hard to reproduce in isolation.

Anyway, new commit incoming. Thanks for reviewing!

Reply:
Yeah, this makes sense, I think the number of follow-on PRs suggests that a revised approach is needed. As you've hinted, perhaps we should initially revert _all_ these PRs and set up a new one with tests that capture all you've learned from the previous PRs? In particular, we probably want to get back to a consistent state before 4.3 goes stable.

> This allows for config for extensions to be overloaded with injector after the extra_methods have been calculated for the original. This was fixing a bug (shown in the unit tests added in that PR) where an overloaded extension was not having its extra methods registered in a key project's unit test suite - it was global state leakage between tests - this came down to the fact that the extra methods for the original Extension class had already been defined in setUpOnce (or similar)

I assume that we're talking specifically about within tests, right? Was the test using $required_extensions, and was that broken? Or was the issue that you wanted to apply the extension for a single test as opposed to a whole test class.

I would recommend that we clearly document the limitations of setting the `extensions` configuration property; this will mitigate the risk of us not being able to make it magically do everything. ;-)

Reply:
>As you've hinted, perhaps we should initially revert all these PRs and set up a new one with tests that capture all you've learned from the previous PRs

Yep totally - this PR now does that (it leaves the existing test that was added to show how overloaded extensions can be used _after_ extra methods are defined for the class it is applied to.

>In particular, we probably want to get back to a consistent state before 4.3 goes stable.

100%

>I assume that we're talking specifically about within tests, right? Was the test using $required_extensions, and was that broken? Or was the issue that you wanted to apply the extension for a single test as opposed to a whole test class.

Yep that's right. The test was _not_ using $required_extensions - the report was that the test was passing in 4.2 and started to break in 4.3. Adding $required_extensions fixed the problem, but the key project was concerned about why it changed since it previously worked, and the extension in question was common in a lot of tests. I've passed a message on to the project that this change is likely to be reverted and that they should use $required_extensions to ensure their tests keep passing.

>It could, for example, be a case of saying "if you've modified the extensions property for a class, you should call SomeClass::someResetStaticMethod()"

Yeah so that's what I'd originally added - `SomeClass::flush_extra_methods_cache()`, but it's ended up causing PHP 5.6 segfaults for some reason (in tests). It's quite a complicated architecture under there, with DataObject using a trait which uses another trait. As you said, we want a consistent state for 4.3 so I think it's safer to revert back to the way it was before and maybe re-assess it as a wider SapphireTest refactor at some point in the future.

We could add some docs saying something like "if you're changing an extension class with Injector after the config manifest has been loaded and after the class you're extending has already been loaded, ensure you call `SomeClass::add_extension(TheExtensionYoureExtending::class)` afterwards to regenerate the extra methods cache" - that's the problem we're trying to fix here

Reply:
It is curious that 4.3 caused a regression here, if it was working in 4.2 - do we know what change from 4.2 originally introduced this problem?

Reply:
I’d like to find some time to investigate for sure